FWCore/Services/plugins/StallMonitor.cc

```diff
@@ -0,0 +1,403 @@
+                                                                           "including stalling-modules information will be logged.");
+  desc.addUntracked<double>("stallThreshold", threshold_default)->setComment("Threshold (in seconds) used to classify modules as stalled.\n"
+                                                                             "Millisecond granularity allowed.");
+  descriptions.add("StallMonitor", desc);
+  descriptions.setComment("This service keeps track of various times in event-processing to determine which modules are stalling.");
+}
+
+void StallMonitor::preModuleConstruction(ModuleDescription const& md)
+{
+  // Module labels are dense, so if the module id is greater than the
+  // size of moduleLabels_, grow the vector to the correct index and
+  // assign the last entry to the desired label.  Note that with the
+  // current implementation, there is no module with ID '0'.  In
+  // principle, the module-information vectors are therefore each one
+  // entry too large.  However, since removing the entry at the front
+  // makes for awkward indexing later on, and since the sizes of these
+  // extra entries are on the order of bytes, we will leave them in
+  // and skip over them later when printing out summaries.  The
+  // extraneous entries can be identified by their module labels being
+  // empty.
+  auto const mid = md.id();
+  if (mid < moduleLabels_.size()) {
+    moduleLabels_[mid] = md.moduleLabel();
+  }
+  else {
+    moduleLabels_.resize(mid+1);
+    moduleLabels_.back() = md.moduleLabel();
+  }
+}
+
+void StallMonitor::postBeginJob()
+{
+  // Since a (push,emplace)_back cannot be called for a vector of a
+  // type containing atomics (like 'StallStatistics')--i.e. atomics
+  // have no copy/move-assignment operators, we must specify the size
+  // of the vector at construction time.
+  moduleStats_ = std::vector<StallStatistics>(moduleLabels_.size());
+  for (std::size_t i{}; i < moduleStats_.size(); ++i) {
+    moduleStats_[i].setLabel(moduleLabels_[i]);
+  }
+
+  if (validFile_) {
+    std::size_t const width {std::to_string(moduleLabels_.size()).size()};
+
+    Column col0 {"Module ID", width};
+    std::string const lastCol {"Module label"};
+
+    std::ostringstream oss;
+    oss << "\n#  " << col0 << space << lastCol << '\n';
+    oss << "#  " << std::string(col0.width()+space.size()+lastCol.size(),'-') << '\n';
+
+    for (std::size_t i{} ; i < moduleLabels_.size(); ++i) {
+      auto const& label = moduleLabels_[i];
+      if (label.empty()) continue; // See comment in filling of moduleLabels_;
+      oss << "#M " << std::setw(width) << std::left << col0(i) << space
+          << std::left << moduleLabels_[i] << '\n';
+    }
+    oss << '\n';
+    file_.write(oss.str());
+  }
+
+  beginTime_ = now();
+}
+
+void StallMonitor::preEvent(StreamContext const& sc)
+{
+  if (!validFile_) return;
```

Comment:
Since you do not want to do anything if `validFile_` is not needed, I'd suggest not even calling `iRegistry.watchPreEvent`. That will actually improve the efficiency of the job a bit since the framework will not have to call this function each event.